File: frontend/src/api/codeGeneration.ts
/**
 * Code Generation API client.
 *
 * Endpoints:
 * - POST /api/code-generation/execute    — execute code task
 * - GET  /api/code-generation/results/{id} — get result
 * - GET  /api/code-generation/results    — list results
 * - POST /api/code-generation/run/python — run Python directly
 * - POST /api/code-generation/run/bash   — run Bash directly
 */
import { client } from './client';

export interface ExecuteCodeRequest {
  task: string;
  context?: string;
  budget?: number;
}

export interface ExecuteCodeResponse {
  task_id: string;
  task_instruction: string;
  completion_reason: string;
  summary: string;
  steps_executed: number;
  budget: number;
  duration_seconds: number;
  execution_history: Array<Record<string, unknown>>;
}

export interface CodeResultResponse {
  task_id: string;
  task_instruction: string;
  completion_reason: string;
  summary: string;
  steps_executed: number;
  duration_seconds: number;
}

export interface CodeResultListItem {
  task_id: string;
  task_instruction: string;
  completion_reason: string;
  steps_executed: number;
}

export interface RunCodeRequest {
  code: string;
  timeout?: number;
}

export interface RunCodeResponse {
  execution_id: string;
  status: string;
  output: string;
  error: string;
  return_code: number;
  duration_seconds: number;
}

export const codeGenerationApi = {
  execute: (body: ExecuteCodeRequest) =>
    client.post<ExecuteCodeResponse>('/api/code-generation/execute', body),

  getResult: (taskId: string) =>
    client.get<CodeResultResponse>(`/api/code-generation/results/${taskId}`),

  listResults: () =>
    client.get<CodeResultListItem[]>('/api/code-generation/results'),

  runPython: (body: RunCodeRequest) =>
    client.post<RunCodeResponse>('/api/code-generation/run/python', body),

  runBash: (body: RunCodeRequest) =>
    client.post<RunCodeResponse>('/api/code-generation/run/bash', body),
};

File: backend/src/skpl_agent/app/_router/_schema/_code_generation.py
"""Request/response schemas for code generation endpoints."""

from __future__ import annotations

from pydantic import BaseModel, Field


class ExecuteCodeRequest(BaseModel):
    task: str = Field(..., min_length=1, description="Task description or code to execute")
    context: str = Field("", description="Additional context")
    budget: int | None = Field(None, ge=1, le=100, description="Max steps")


class ExecuteCodeResponse(BaseModel):
    task_id: str
    task_instruction: str
    completion_reason: str
    summary: str
    steps_executed: int
    budget: int
    duration_seconds: float
    execution_history: list[dict]


class CodeResultResponse(BaseModel):
    task_id: str
    task_instruction: str
    completion_reason: str
    summary: str
    steps_executed: int
    duration_seconds: float


class CodeResultListItem(BaseModel):
    task_id: str
    task_instruction: str
    completion_reason: str
    steps_executed: int


class RunPythonRequest(BaseModel):
    code: str = Field(..., min_length=1, description="Python code to execute")
    timeout: int = Field(30, ge=1, le=300, description="Timeout in seconds")


class RunBashRequest(BaseModel):
    code: str = Field(..., min_length=1, description="Bash code to execute")
    timeout: int = Field(30, ge=1, le=300, description="Timeout in seconds")


class RunCodeResponse(BaseModel):
    execution_id: str
    status: str
    output: str
    error: str
    return_code: int
    duration_seconds: float